File: 133-Clone-Graph/DFS.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {
    public Node cloneGraph(Node node) {
        if (node == null) return null;
        
        HashMap<Integer, Node> cloned = new HashMap<>();
        cloned.put(node.val, new Node(node.val));
        
        Stack<Node> stack = new Stack<>();
        stack.push(node);
        
        while (!stack.isEmpty()) {
            Node crr = stack.pop();
            for (Node adj : crr.neighbors) {
                if (!cloned.containsKey(adj.val)) {
                    cloned.put(adj.val, new Node(adj.val));
                    stack.push(adj);
                }
                cloned.get(crr.val).neighbors.add(cloned.get(adj.val));
            }
        }
        
        return cloned.get(node.val);
    }
}

File: 199-Binary-Tree-Right-Side-View/DFS.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun rightSideView(root: TreeNode?): List<Int> {
        val values = mutableListOf<Int>()
        if (root == null) {
            return values
        }

        val stack = ArrayDeque<Pair<Int, TreeNode>>();
        stack.push(0 to root)
        
        while (!stack.isEmpty()) {
            val (depth, node) = stack.pop()
            if (values.size == depth) {
                values.add(node.`val`)
            }
            
            if (node.left != null) {
                stack.push(depth + 1 to node.left)
            }
            
            if (node.right != null) {
                stack.push(depth + 1 to node.right)
            }
        }
        
        return values
        
    }
}
